### ISSUE TITLE
Incremental builds failing

### ISSUE THREAD

**Issue by weirddan455** [dosbox-staging/dosbox-staging#3693]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Linux x86_64

### If Other OS, please describe

Debian 12

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

I keep having this issue when running bisects.  Sometimes the incremental build will fail because it thinks standard C++ types and methods are undefined.  In this case it was semaphore but I've also seen this happen with `std::string` methods.

This error repeats itself for multiple files.  Deleting the build directory and doing a clean build resolves the issue.  I know we use pre-compiled headers in some places and I'm wondering if those are at fault.

I am not using ccache so that is not to blame.

### Steps to reproduce the behaviour.

Do a bisect from v0.80.1 to main.  Some incremental builds can fail.  I can try to find 2 commits that reproduce this issue later.  I'm in the middle of working on something else right now but wanted to go ahead and raise an issue in case anyone has ideas.

### Provide a Log

```text
[41/289] Compiling C++ object src/libs/decoders/libdecoders.a.p/opus.cpp.o
FAILED: src/libs/decoders/libdecoders.a.p/opus.cpp.o 
clang++-16 -Isrc/libs/decoders/libdecoders.a.p -Isrc/libs/decoders -I../../src/libs/decoders -I../../include -I. -I../.. -Isrc/libs/ghc -I../../src/libs/ghc -Isrc/libs/loguru -I../../src/libs/loguru -fcolor-diagnostics -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -std=c++20 -O0 -g -Wno-unknown-pragmas -fpch-preprocess -Weffc++ -Wextra-semi -D_GLIBCXX_ASSERTIONS=1 -D_GLIBCXX_DEBUG=1 -D_GLIBCXX_DEBUG_PEDANTIC=1 -D_GLIBCXX_SANITIZE_VECTOR=1 -D_LIBCPP_DEBUG=1 -fPIC -DLOGURU_STACKTRACES=1 -DLOGURU_USE_LOCALE=0 -isystem/usr/include/opus -isystem/usr/include/SDL2 -D_REENTRANT -isystem../../src/libs -isystemsrc/libs -MD -MQ src/libs/decoders/libdecoders.a.p/opus.cpp.o -MF src/libs/decoders/libdecoders.a.p/opus.cpp.o.d -o src/libs/decoders/libdecoders.a.p/opus.cpp.o -c ../../src/libs/decoders/opus.cpp
In file included from ../../src/libs/decoders/opus.cpp:46:
In file included from ../../include/math_utils.h:27:
In file included from ../../include/support.h:42:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/thread:40:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/stop_token:37:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/semaphore:35:
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/semaphore_base.h:180:5: error: unknown type name 'sem_t'
    sem_t _M_semaphore;
    ^
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by weirddan455** [comment]
I can reproduce this now by doing a clean build of main currently commit 407a9222270ef2415d87a4daa72c51d3ed86d1dd then checking out v0.81.1 tag and attempting an incremental build.  The reverse also fails (clean build of v0.81.1 followed by attempting incremental build of main).

**Comment by interloper98** [comment]
Is it a boundary when the project changed C++ standards? 

Whenever I've compiled new releases that changed the C++ standard (or changed dependencies quite a bit), I had to reconfigure/re-setup/(or `git clean -ffdx`).

**Comment by weirddan455** [comment]
That might be it, yeah.  Meson is supposed to automatically reconfigure itself though.  I see it do that often with incremental builds.  I just delete the build directory and do a clean build when I run into this bug and then it's fine.

It's possible this is just a Meson bug (and I'm on Debian so I also have an older version of Meson).  Not sure if there's much we can do about this.  We're going to be switching to CMake anyway.

**Comment by interloper98** [comment]
> Meson is supposed to automatically reconfigure itself though.

Argh; I didn't know that. That's too bad it can't catch this! 

I would think something as simple as a checksum compare on all the meson project files could do the job.  If anything's changed, just be safe and re-configure the ground up.  I guess some people would complain _"I made an inconsequential change, Meson is wasting my time re-configuring when it doesn't need to"_; but hey - I'd rather it get these right all the time.

**Comment by dreamer** [comment]
The question is: if this is something worth fixing? Doing clean build when changing C++ standard seems like a reasonable thing to do...

**Comment by johnnovak** [comment]
Do we need this to be open still? I don't think there is a real issue here, plus the switch to CMake will make this irrelevant.

We'll, kinda, as we'll still have to use Meson when bisecting into pre-CMake times...

**Comment by weirddan455** [comment]
I don't know that there is a fix here.  It's probably a Meson bug not triggering a clean rebuild when it should so we just have to do that manually when bisecting past when certain settings have been changed.

I created this ticket because I got annoyed when doing bisects to see if anyone has any ideas.  It doesn't sound like there's anything actionable though so we can close this.

**Comment by interloper98** [comment]
If we can detect a language change in our own source files (something like config.h that's generated/populated), then we could error out with a nice message. This could work regardless of what build system were using.

Will try to add this provided it's not too ugly.

**Comment by interloper98** [comment]
(assigning to me; my hopenof success is slim 😆 and won't waste much time on it. )

**Comment by weirddan455** [comment]
Well if you want to spend time on this @interloper98 that's fine I suppose.  Before starting on this, I would check that you can reproduce this on your own machine.  I'm using the Meson package from Debian which has an outdated Meson version 1.0.1.  If this is indeed a Meson bug, it's possible it got fixed in a newer version.

Also, if your fix is just to throw a nice error message, I would think that would make a non-buggy build system that can auto-detect and rebuild fail for no good reason.  And a buggy build system would still require a manual clean build.

**Comment by Grounded0** [comment]
Outdated Debian package? That can't be possible. 😁

**Comment by interloper98** [comment]
> Meson package from Debian which has an outdated Meson version 1.0.1

Woah; it's 15 months stale! :cheese:

I'm on retropie for my build environment (which is a pretty old debian), and had to use `pip` to get Meson. It's kept up to date, and no problems. https://pypi.org/project/meson/

**Comment by interloper98** [comment]
@weirddan455, well I wasn't able to flag it using macros (I tried writing `__cplusplus` into `config.h` as a literal during `meson setup` and compared that  against the compiler's live `__cplusplus` macro, thinking I could flag the discrepancy .. but, they always matched :laughing:).  So Meson is internally consistent between the setup and compile phases.

But some good news is that the `--wipe` argument can be added during setup: `meson setup --wipe build`, and can be run before building to guarantee the build area gets the correct setting.

Basically the the `setup --wipe` runs about as fast as when meson detects a configuration change, anyway, so for bisecting a build alias like: `meson setup --wipe build && meson compile -C build` will always work.

I can flip back-and-fourth between older git tagged releases and back up to `main`, and it now always builds correctly.

**Comment by johnnovak** [comment]
> The `setup --wipe` runs about as fast as when meson detects a configuration change, anyway, so for bisecting a build alias like: `meson setup --wipe build && meson compile -C build` will always work.
> 
> I can flip back-and-fourth between older git tagged releases and back up to `main`, and it now always builds correctly.

Could you add these tips to the end of https://github.com/dosbox-staging/dosbox-staging/blob/main/BUILD.md in a new "Bisecting tips" section or something?

As I mentioned, Meson won't go completely away for a year or two at least if we need to bisect far enough into the history.

**Comment by interloper98** [comment]
Will add that :+1:
